internal/primitives/core/crypto/crypto.go

```diff
@@ -0,0 +1,246 @@
+// generally be equivalent to no password at all.
+//
+// # Example
+//
+// Parse [`DEV_PHRASE`] secret uri with junction:
+//
+// ```
+// # use sp_core::crypto::{SecretUri, DeriveJunction, DEV_PHRASE, ExposeSecret};
+// # use std::str::FromStr;
+// let suri = SecretUri::from_str("//Alice").expect("Parse SURI");
+//
+// assert_eq!(vec![DeriveJunction::from("Alice").harden()], suri.junctions);
+// assert_eq!(DEV_PHRASE, suri.phrase.expose_secret());
+// assert!(suri.password.is_none());
+// ```
+//
+// Parse [`DEV_PHRASE`] secret ui with junction and password:
+//
+// ```
+// # use sp_core::crypto::{SecretUri, DeriveJunction, DEV_PHRASE, ExposeSecret};
+// # use std::str::FromStr;
+// let suri = SecretUri::from_str("//Alice///SECRET_PASSWORD").expect("Parse SURI");
+//
+// assert_eq!(vec![DeriveJunction::from("Alice").harden()], suri.junctions);
+// assert_eq!(DEV_PHRASE, suri.phrase.expose_secret());
+// assert_eq!("SECRET_PASSWORD", suri.password.unwrap().expose_secret());
+// ```
+//
+// Parse [`DEV_PHRASE`] secret ui with hex phrase and junction:
+//
+// ```
+// # use sp_core::crypto::{SecretUri, DeriveJunction, DEV_PHRASE, ExposeSecret};
+// # use std::str::FromStr;
+// let suri = SecretUri::from_str("0xe5be9a5092b81bca64be81d212e7f2f9eba183bb7a90954f7b76361f6edb5c0a//Alice").expect("Parse SURI");
+//
+// assert_eq!(vec![DeriveJunction::from("Alice").harden()], suri.junctions);
+// assert_eq!("0xe5be9a5092b81bca64be81d212e7f2f9eba183bb7a90954f7b76361f6edb5c0a", suri.phrase.expose_secret());
+// assert!(suri.password.is_none());
+// ```
+type SecretURI struct {
+	// The phrase to derive the private key.
+	// This can either be a 64-bit hex string or a BIP-39 key phrase.
+	Phrase string
+	// Optional password as given as part of the uri.
+	Password *string
+	// The junctions as part of the uri.
+	Junctions []DeriveJunction
+}
+
+func NewSecretURI(s string) (SecretURI, error) {
+	matches := secretPhraseRegex.FindStringSubmatch(s)
+	if matches == nil {
+		return SecretURI{}, fmt.Errorf("invalid format")
+	}
+
+	var (
+		junctions []DeriveJunction
+		phrase    = DevPhrase
+		password  *string
+	)
+	for i, name := range secretPhraseRegex.SubexpNames() {
+		if i == 0 {
+			continue
+		}
+		switch name {
+		case "path":
+			junctionMatches := junctionRegex.FindAllString(matches[i], -1)
+			for _, jm := range junctionMatches {
+				junctions = append(junctions, NewDeriveJunctionFromString(jm))
+			}
+		case "phrase":
+			if matches[i] != "" {
+				phrase = matches[i]
+			}
+		case "password":
+			if matches[i] != "" {
+				pw := matches[i]
+				password = &pw
+			}
+		}
+	}
+	return SecretURI{
+		Phrase:    phrase,
+		Password:  password,
+		Junctions: junctions,
+	}, nil
+}
+
+// Trait suitable for typical cryptographic PKI key pair type.
+//
+// For now it just specifies how to create a key from a phrase and derivation path.
+type Pair[Seed, Signature any] interface {
```

Comment:
Im a little confused what this file is actually for. We have implementations of the different crypto protocols that we need already, so if this meant to just wrap them/replace them? Its unclear to me what the purpose of this is. Could you elaborate on how how this gets used? With the regexes and such it looks quite different to me from our existing impls and I'm curious why that is

Reply:
When trying to debug the tests after trying to fix signing, I found it difficult to know if the encodings were correct based on our existing keyring.  I discovered they used `SecretUri` to derive the keys using specific syntax.  Given that this syntax is synonymous in substrate, I thought it prudent to recreate how the keys were derived rather than change our existing keyring impl.  This would be useful to support the substrate like syntax when they declare authorities like `\\Alice` from the command line or chain spec.